pkg/logstructured/logstructured.go

```diff
@@ -294,28 +296,62 @@ func (l *LogStructured) ttlEvents(ctx context.Context) chan *server.Event {
 				}
 			}
 		}
+
+		if ctx.Err() == nil {
+			logrus.Error("ttl events watch channel was closed due to slow processing")
+		}
 	}()
 
 	return result
 }
 
 func (l *LogStructured) ttl(ctx context.Context) {
 	// vary naive TTL support
-	mutex := &sync.Mutex{}
-	for event := range l.ttlEvents(ctx) {
-		go func(event *server.Event) {
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(time.Duration(event.KV.Lease) * time.Second):
-			}
-			mutex.Lock()
-			if _, _, _, err := l.Delete(ctx, event.KV.Key, event.KV.ModRevision); err != nil {
-				logrus.Errorf("failed to delete expired key: %v", err)
-			}
-			mutex.Unlock()
-		}(event)
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter())
```

Comment:
I don't think we want to put a rate limit on things expiring, do we? Would `NewDelayingQueue()` be more appropriate?

Reply:
NewRateLimitingQueue support delaying queue with method queue.AddAfter and offer method queue.AddRateLimited

Reply:
My question was why we want to rate limit at all. Things should be deleted as they expire. I do not understand why we would want to rate limit the timely removal of deleted items. Can you explain why that would be desired?

Reply:
Due to RateLimitingQueue support all functions of delaying queue.  event is put into queue with fixed event lease interval. **When event error deleting, here just use it rate limit function, the rate limit function provide exponential retries**. For the error retrying, Maybe, we also can use delayingQueue with fixed interval.

`// NewRateLimitingQueue constructs a new workqueue with rateLimited queuing ability
// Remember to call Forget!  If you don't, you may end up tracking failures forever.
// NewRateLimitingQueue does not emit metrics. For use with a MetricsProvider, please use
// NewNamedRateLimitingQueue instead.
func NewRateLimitingQueue(rateLimiter RateLimiter) RateLimitingInterface {
	return &rateLimitingType{
		DelayingInterface: NewDelayingQueue(),
		rateLimiter:       rateLimiter,
	}
}
`

Reply:
You're not making a case for using exponential retries or rate limiting of queue entries. Why do you want that?  I don't think those are things we want. I think a simple delaying queue is more appropriate. Can you explain why that is not the case?

Reply:
I got it. I will replace the RateLimitingQueue with DelayingQueue later.

Reply:
> You're not making a case for using exponential retries or rate limiting of queue entry processing. Why do you want that? I don't think those are things we want. I think a simple delaying queue is more appropriate. Can you explain why that is not the case?

Fixed is done.